### PR TITLE
[BESU-163] Add inbound and outbound TLS information.

### PR DESCRIPTION
**Pull request by bgravenorst**
Signed-off-by: Byron Gravenorst <byron.gravenorst@consensys.net>

## PR description
Add information about configuring inbound TLS in Besu.

## Fixed Issue(s)
BESU-163

### CODE REVIEW

File: docs/Concepts/TLS.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+description: TLS overview
+<!--- END of page meta data -->
+
+# TLS Communication
+
+Hyperledger Besu supports inbound TLS communication (for example from EthSigner
+or curl).
+
+All TLS versions are supported. However, certificates must be PKCS #12
```
Comment (usmansaleem) [inline docs/Concepts/TLS.md]: This is slightly misleading. We support TLSv1, TLSv1.1 and TLSv1.2. We don't yet support TLSv1.3 (I don't know whether we should only mention the SSL/TLS versions we support and/or also mention the versions we don't support).

Private Keys and Certificates are stored in a PKCS#12 format file which is password protected.

File: docs/HowTo/Configure/Configure-TLS.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+description: Configure TLS
```
Comment (usmansaleem) [inline docs/HowTo/Configure/Configure-TLS.md]: We are specifically talking about configuring TLS for JSON-RPC HTTP endpoint. This feels a bit more generic, isn't it? i.e. Besu can and will support TLS for multiple endpoints in future release both in-bound and out-bound.

File: docs/HowTo/Configure/Configure-TLS.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+**Prerequisites**:
+
+* Password-protected PKSC #12 keystore.
+* File containing the keystore password
+
+## Create Known Clients File
+
+Create a file (in this example, `knownClients`) that lists one or more clients
+with permission to connect to Besu. Use the format`<common_name> <hex-string>`
```
Comment (usmansaleem) [inline docs/HowTo/Configure/Configure-TLS.md]: This is also slightly misleading (and we might change its meaning in a future commit as well), basically we are enabling TLS "client authentication" on our endpoint. The "knownClients" file, at the moment, can be used to record self-signed client certificates fingerprint. At the moment, this file is **NOT** being used as permissions to connect to Besu (we will be bringing in a change in Besu code very soon which will make this file as a 'permission' file - but for now, this is not a permission file).

File: docs/Reference/CLI/CLI-Syntax.md [hunk 4]
```diff
@@ -1517,6 +1517,93 @@ rpc-http-port="3435"
...
 The default is 8545. Ports must be [exposed appropriately](../../HowTo/Find-and-Connect/Configuring-Ports.md). 
 
+### rpc-http-tls-enabled
+
+```bash tab="Syntax"
+--rpc-http-tls-enabled
+```
+
+```bash tab="Environement Variable"
+BESU_RPC_HTTP_TLS_ENABLED=true
```
Comment (usmansaleem) [inline docs/Reference/CLI/CLI-Syntax.md]: I need to check this, have you test that this env variable works as expected?

File: docs/Reference/CLI/CLI-Syntax.md [hunk 5]
```diff
@@ -1517,6 +1517,93 @@ rpc-http-port="3435"
...
+```
+
+Enables client authentication.
+
+For client certificates not signed by a known root CA this must contain the
+certifcates's Common Name, and SHA-256 fingerprint in the format
+`<CommonName> <hex-string>`.
+
+Client certificates signed by a known CA do not need to be added to the file.
+
```
Comment (usmansaleem) [inline docs/Reference/CLI/CLI-Syntax.md]: This is the bit that will be changed in an upcoming commit. We are going to introduce client-auth disabled, CA-ONLY and WHITELIST_ONLY mode. The "knownClients" will become whitelistClients file or something on similar lines.

Comment (usmansaleem) [reply docs/Reference/CLI/CLI-Syntax.md]: i.e. we don't want empty file to enable client-auth that we have right now.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 4 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: @usmansaleem, no I have not tested this. Will give it try when I test this feature again.

File: docs/Concepts/TLS.md (hunk 1 above)
Comment (bgravenorst) [reply docs/Concepts/TLS.md]: I think it's probably best to mention which version is not supported.

File: docs/Concepts/TLS.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+description: TLS overview
+<!--- END of page meta data -->
+
+# TLS Communication
+
+Hyperledger Besu supports TLS to secure client and server communication. The
+client (EthSigner) or server (Orion) must also be configured for TLS
```
[the remark this answers is not in the document]
Comment (MadelineMurray) [reply docs/Concepts/TLS.md]: We say Orion throughout the docs already so I think it's OK.  There's a note in the concept content saying to use privacy you need a private transaction manager and Orion is the example used in these docs

File: docs/HowTo/Configure/Configure-TLS.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+# Configure TLS
+
+Hyperledger Besu supports TLS for client and server communication. For example, you can
+[configure TLS](../../Concepts/TLS.md) for communication between EthSigner and Besu, and Besu and Orion.
+
+Configure TLS communication from the command line.
+
+**Prerequisites**:
+
+* Besu's password-protected PKSC #12 keystore.
```
[the remark this answers is not in the document]
Comment (MadelineMurray) [reply docs/HowTo/Configure/Configure-TLS.md]: Does this change need made in the EthSigner PR as well?

File: docs/HowTo/Configure/Configure-TLS.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+<!--- END of page meta data -->
+
+# Configure TLS
+
+Hyperledger Besu supports TLS for client and server communication. For example, you can
+[configure TLS](../../Concepts/TLS.md) for communication between EthSigner and Besu, and Besu and Orion.
+
+Configure TLS communication from the command line.
+
+**Prerequisites**:
```
Comment (MadelineMurray) [inline docs/HowTo/Configure/Configure-TLS.md]: Similar to comment on EthSigner PR - I think it's worth making it clearer these prereqs are common to client and server.  The title could be changed to reflect this: 
_Prerequisites for Client and Server TLS_

File: docs/HowTo/Configure/Configure-TLS.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+* `<common_name>` is the Common Name specified in the client certificate.
+* `<hex-string>` is the SHA-256 fingerprint of the client certificate.
+
+!!! example
+    ```
+    ethsigner 8E:E0:85:9F:FC:2E:2F:21:31:46:0B:82:4C:A6:88:AB:30:34:9A:C6:EA:4F:04:31:ED:0F:69:A7:B5:C2:2F:A7
+    curl FC:18:BF:39:45:45:9A:15:46:76:A6:E7:C3:94:64:B8:34:84:A3:8E:B8:EA:67:DC:61:C0:29:E6:38:B8:B7:99
+    ```
+    
+You can use `openssl` or `keytool` to display the SHA256 fingerprint.
```
Comment (MadelineMurray) [inline docs/HowTo/Configure/Configure-TLS.md]: Are there links we could include for openssl and keytool?

File: docs/HowTo/Configure/Configure-TLS.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+The command line:
+
+* Enables the HTTP JSON-RPC service using the [`--rpc-http-enabled`](../../Reference/CLI/CLI-Syntax.md#rpc-http-enabled) option.
+* Enables TLS for the HTTP JSON-RPC service using the [`--rpc-http-tls-enabled`](../../Reference/CLI/CLI-Syntax.md#rpc-http-tls-enabled)
+option.
+* Enables TLS client authentication using the [`--rpc-http-tls-client-auth-enabled`](../../Reference/CLI/CLI-Syntax.md#rpc-http-tls-client-auth-enabled) option.
+* Specifies the keystore using the [`--rpc-http-tls-keystore-file`](../../Reference/CLI/CLI-Syntax.md#rpc-http-tls-keystore-file) option.
+* Specifies the file that contains the password to decrypt the keystore using
+the [`--rpc-http-tls-keystore-password-file`](../../Reference/CLI/CLI-Syntax.md#rpc-http-tls-keystore-password-file) option.
+* [Specify the clients](#create-the-known-clients-file) that are allowed to connect to Besu using the
```
Comment (MadelineMurray) [inline docs/HowTo/Configure/Configure-TLS.md]: Specify -> Specifies

File: docs/HowTo/Configure/Configure-TLS.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+[`--rpc-http-tls-known-clients-file`](../../Reference/CLI/CLI-Syntax.md#rpc-http-tls-known-clients-file) option.
+
+!!! note
+    Set [`--rpc-http-tls-ca-clients-enabled`](../../Reference/CLI/CLI-Syntax.md#rpc-http-tls-ca-clients-enabled)
+    to `true` to allow access to clients with signed and trusted root CAs.
+    
+## Configure Server TLS
+
+Allow Besu to securely communicate with the server (Orion)
```
Comment (MadelineMurray) [inline docs/HowTo/Configure/Configure-TLS.md]: Needs a fullstop

File: docs/HowTo/Configure/Configure-TLS.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+* `<hex-string>` is the SHA-256 fingerprint of the server's certificate.
+
+!!! example
+    ```
+    localhost:8888 3C:B4:5A:F9:88:43:5E:62:69:9F:A9:9D:41:14:03:BA:83:24:AC:04:CE:BD:92:49:1B:8D:B2:A4:86:39:4C:AC
+    127.0.0.1:8888 3C:B4:5A:F9:88:43:5E:62:69:9F:A9:9D:41:14:03:BA:83:24:AC:04:CE:BD:92:49:1B:8D:B2:A4:86:39:4C:AC
+    ```
+
+!!! note
+    Specify both hostname and IP address in the file if unsure which is used in
```
Comment (MadelineMurray) [inline docs/HowTo/Configure/Configure-TLS.md]: which is -> which one is?

File: docs/Reference/CLI/CLI-Syntax.md [hunk 13]
```diff
@@ -1180,6 +1180,92 @@ privacy-public-key-file="Orion/nodeKey.pub"
...
+```bash tab="Environment Variable"
+BESU_PRIVACY_TLS_ENABLED=false
+```
+
+```bash tab="Configuration File"
+privacy-tls-enabled=false
+```
+
+Enables [TLS with the Private Transaction Manager](../../Concepts/Privacy/Privacy-Overview.md#private-transaction-manager).
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: TLS with the -> TLS on communication with the?

File: docs/Reference/CLI/CLI-Syntax.md [hunk 14]
```diff
@@ -1180,6 +1180,92 @@ privacy-public-key-file="Orion/nodeKey.pub"
...
+```bash tab="Environment Variable"
+BESU_PRIVACY_TLS_KNOWN_ENCLAVE_FILE=/home/me/me_node/knownEnclave
+```
+
+```bash tab="Configuration File"
+privacy-tls-known-enclave-file="/home/me/me_node/knownEnclave"
+```
+
+Path to a file containing the hostnames, ports, and SHA256 certificate fingerprints
+of the [authorized privacy enclave](../../HowTo/Configure/Configure-TLS.md##create-the-known-servers-file).
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: Extra # in link

File: docs/Reference/CLI/CLI-Syntax.md [hunk 15]
```diff
@@ -1517,6 +1603,133 @@ rpc-http-port="3435"
 Specifies HTTP JSON-RPC listening port (TCP).
 The default is 8545. Ports must be [exposed appropriately](../../HowTo/Find-and-Connect/Configuring-Ports.md). 
 
+### rpc-http-tls-ca-clients-enabled
+
+```bash tab="Syntax"
+--rpc-http-tls-ca-clients-enabled
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: The syntax is:  
`--rpc-http-tls-ca-clients-enabled[=<true|false>]`

File: docs/Reference/CLI/CLI-Syntax.md [hunk 16]
```diff
@@ -1517,6 +1603,133 @@ rpc-http-port="3435"
...
+```
+
+```bash tab="Environment Variable"
+BESU_RPC_HTTP_TLS_CA_CLIENTS_ENABLED[=<true|false>]
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: To be consistent, I think this needs to be an example: 
`BESU_RPC_HTTP_TLS_CA_CLIENTS_ENABLED=true`

File: docs/Reference/CLI/CLI-Syntax.md [hunk 17]
```diff
@@ -1517,6 +1603,133 @@ rpc-http-port="3435"
...
+
+```bash tab="Environment Variable"
+BESU_RPC_HTTP_TLS_KNOWN_CLIENTS_FILE=/home/me/me_node/knownClients
+```
+
+```bash tab="Configuration File"
+rpc-http-tls-known-clients-file="/home/me/me_node/knownClients"
+```
+
+Used to [authenticate clients](../../HowTo/Configure/Configure-TLS.md##create-the-known-clients-file) that use self-signed certificates or non-public
```
Comment (MadelineMurray) [inline docs/Reference/CLI/CLI-Syntax.md]: Extra # in link

File: docs/HowTo/Configure/Configure-TLS.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/HowTo/Configure/Configure-TLS.md]: Done

File: docs/HowTo/Configure/Configure-TLS.md [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+## Configure Client TLS
+
+Allow clients (for example a dApp, curl, or EthSigner) to send and receive
+secure HTTP JSON-RPCs.
+
+**Prerequisites**:
+
+* The client must be configured for TLS.
+* Client's PKSC #12 keystore information.
```
[the remark this answers is not in the document]
Comment (bgravenorst) [reply docs/HowTo/Configure/Configure-TLS.md]: Done

File: docs/HowTo/Configure/Configure-TLS.md (hunk 8 above)
Comment (bgravenorst) [reply docs/HowTo/Configure/Configure-TLS.md]: Done

File: docs/HowTo/Configure/Configure-TLS.md (hunk 10 above)
Comment (bgravenorst) [reply docs/HowTo/Configure/Configure-TLS.md]: Done

File: docs/HowTo/Configure/Configure-TLS.md (hunk 9 above)
Comment (bgravenorst) [reply docs/HowTo/Configure/Configure-TLS.md]: Done

File: docs/HowTo/Configure/Configure-TLS.md (hunk 11 above)
Comment (bgravenorst) [reply docs/HowTo/Configure/Configure-TLS.md]: Done

File: docs/Reference/CLI/CLI-Syntax.md (hunk 13 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done

File: docs/Reference/CLI/CLI-Syntax.md (hunk 14 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done

File: docs/Reference/CLI/CLI-Syntax.md (hunk 15 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 16 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done. Mistakenly copied syntax example to Env Variable section.

File: docs/Reference/CLI/CLI-Syntax.md (hunk 17 above)
Comment (bgravenorst) [reply docs/Reference/CLI/CLI-Syntax.md]: Done

**Pull request merged**